### DISCUSSION TITLE
toggle preview between selection and file preview

### DISCUSSION THREAD

**Discussion by rbusfield** [junegunn/fzf#3574] [Q&A] [answered]
Original post was rather wide with 3 questions.  
The final Answer is regarding ability to toggle preview.  I will repost a new discussion with the fzf.vim question.

Editing the original post:
> New to fzf.vim and vim plugins in general...  
> I have a config file (yml-like) that I edit in vim.  One section of the config file takes items from a huge list of items (thousands).
> Ideally, a custom vim command binding (:MyCmd) would run my list generator and feed list into fzf.  The selected items should be inserted into the config file with preceding spaces.
>     foo
>     bar
>     baz
> similar to a `:r! cat list.txt | grep -e "foo" -e "ba*" | sed -i -e 's/^/    /'`
> 
> 
> **(answered)** Extra points for showing the current selection of items in a preview window.
> **(answered)** Even more points for ability to toggle between custom item preview and current selection preview.

**Comment by LangLangBart** [comment]
> Even more points for ability to toggle between custom item preview and current selection preview.

One potential solution involves defining two hotkeys to display the preferred preview. For instance,

```bash
seq 12 | fzf --multi --preview 'echo {+}' --preview-window 'nohidden' \
    --bind 'load:select-all' \
    --bind 'ctrl-t:change-preview:echo {+}' \
    --bind 'ctrl-g:change-preview:cat {+f}'
```

<img src="https://github.com/junegunn/fzf/assets/92653266/207c43f4-b7fc-4aed-91e9-a01e5d0d703e" width="400">

---

In the most recent `fzf` version[^1], a `transform` action was introduced that enables the toggling
between two previews with a single hotkey[^2]. 

```bash
seq 12 | fzf --multi --preview 'echo {+}' --preview-window 'nohidden' \
    --bind 'load:select-all' \
    --prompt 'First Preview > ' \
    --bind 'ctrl-t:transform:[[ ! {fzf:prompt} =~ First ]] &&
        echo "change-prompt(First Preview > )+change-preview:echo {+}" ||
        echo "change-prompt(Second Preview > )+change-preview:echo {}"'
```

>[!WARNING]
> ~~Please be aware that there is currently a problem[^3] with the `{+f}` placeholder in the `transform`
> action. And, the preview will not update during a focus[^4] shift (<kbd>↑/↓</kbd>)~~.
> The `{}` placeholder needs to be escaped in the above example.

>```diff
>        --bind 'ctrl-t:transform:[[ ! {fzf:prompt} =~ First ]] &&
>-        echo "change-prompt(First Preview > )+change-preview:echo {+}" ||
>-        echo "change-prompt(Second Preview > )+change-preview:echo {}"'
>+        echo "change-prompt(First Preview > )+change-preview:echo \{+}" ||
>+        echo "change-prompt(Second Preview > )+change-preview:echo \{}"'
>```


[^1]: [CHANGELOG.md 0.45.0](https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0450)
[^2]: [ADVANCED.md Toggling with a single key binding](https://github.com/junegunn/fzf/blob/master/ADVANCED.md#toggling-with-a-single-key-binding)
[^3]: [Issue #3575](https://github.com/junegunn/fzf/issues/3575)
[^4]: [Issue #3579](https://github.com/junegunn/fzf/issues/3579)

**Comment by rbusfield** [comment]
HMMM.  I can see how this works but think I am still stuck with the two Issues above.  
Here is my testcase script `fzf-transform.sh` but still unsure of the escape for the preview action inside the transform.:

```
#!/bin/bash
# fzf testcase for Preview transform/toggle between Selected_Files and File_Preview
# Requirements:
#   1.1 One input list without reload
#   1.2 preview on current item (cat {})
#   1.3 toggle to preview of the selected list of items, one per line (echo {+f})
#   1.4 toggle between current item and selected set (CTRL-T)
#   1.5 retain selected item list between transforms
#   1.6 return seleted list as stdout

#create list
dir=`mktemp -d`
cat << EOF > $dir/list.txt
foo
bar
baz
EOF

# create files
echo "# file is foo" > $dir/foo
echo "# file is bar" > $dir/bar
echo "# file is baz" > $dir/baz

pushd $dir
cat list.txt | \
fzf --multi --prompt 'Items> ' \
  --header 'CTRL-T: Toggle Preview, [TAB]: multi-select' \
  --preview-label ' File_Preview ' \
  --bind 'ctrl-t:transform:[[ ! {fzf:preview-label} =~ Selected_Files ]] &&
      echo "change-preview-label(" Selected_Files ")+change-preview:echo \{+f}" ||
      echo "change-preview-label(" File_Preview ")+change-preview:cat \{}"' \
  --preview '[[ {fzf:preview-label} =~ File_Preview ]] && cat {} || echo {+}'

popd
rm -rf $dir
```

**Reply by LangLangBart** [reply, marked as answer]
As of `fzf` version `0.45.0`, the `{fzf:preview-label}` is not a valid placeholder.


|     Placeholder      |            Description            |
| :------------------: | :-------------------------------: |
|         `{}`         |           current line            |
|        `{n}`         | current line number (zero index)  |
| `{q}` /`{fzf:query}` |           current query           |
|    `{fzf:prompt}`    |          current prompt           |
|    `{fzf:action}`    | name of the last action performed |


| Placeholder expression flags |                Description                 |
| :--------------------------: | :----------------------------------------: |
|             `+`              | space-separated list of the selected lines |
|             `f`              |               temporary file               |
|             `s`              |            preserve whitespace             |

The `{fzf:prompt}` was added so that users could use it as a state[^1], instead of creating an external file. There's also a discussion about changing the `{fzf:prompt}` to `$FZF_PROMPT` in the next version[^2]. If you prefer not to use the `{fzf:prompt}`, you might need to use an external file to manage your state, like so:

```bash
# File_Preview: state 0
# Selected_Files: state 1
echo 0 >/tmp/state
cat list.txt | fzf --multi --prompt 'Items> ' \
	--header 'CTRL-T: Toggle Preview, [TAB]: multi-select' \
	--preview-label ' File_Preview ' \
	--preview 'cat {}' \
	--bind 'ctrl-t:transform:[[ $(</tmp/state) -eq 0 ]] &&
		echo "execute-silent(echo 1 > /tmp/state)+change-preview-label( Selected_Files )+change-preview:cat \{+f}" ||
		echo "execute-silent(echo 0 > /tmp/state)+change-preview-label( File_Preview )+change-preview:cat \{}"'
```

<img src="https://github.com/junegunn/fzf/assets/92653266/1f839211-ad5a-45c6-ba1d-614f8867ffa9" width="500">


[^1]: [Possibility to access the current prompt through a placeholder · Issue #3354](https://github.com/junegunn/fzf/issues/3354)
[^2]: [Expose states to the subprocesses via environment variables · Issue #3582 ](https://github.com/junegunn/fzf/issues/3582)

**Reply by rbusfield** [reply]
Changing the Subject line to reflect the solution for toggle preview
